### PR TITLE
5 tips for a smoother code review process

### PR DESCRIPTION
**Pull request by michaelcarter**
### Blog title
5 tips for a smoother code review process

### Blog description
Title is fairly self explanatory, details some of the customisations we've made to our code review process for efficiency.

### Other notes
Please don't merge this at the moment, the links to the culture repo are broken because it's private. I'm going to chat to Rog about opening this up as there's no reason for it not to be public right now.

#### Author
- [x] I have checked for grammatical errors
- [x] I have checked for spelling errors
- [x] All information (technical or otherwise) is correct to my knowledge

#### Editor 1
- [x] I have checked for grammatical errors
- [x] I have checked for spelling errors
- [x] All information (technical or otherwise) is correct to my knowledge

#### Editor 2
- [x] I have checked for grammatical errors
- [x] I have checked for spelling errors
- [x] All information (technical or otherwise) is correct to my knowledge

### CODE REVIEW

File: _posts/2015-03-18-5-tips-for-a-smoother-code-review-process.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+comments: "true"
+layout: post
+published: true
+title: "5 tips for a smoother code review process"
+author: mike_carter
+categories: ""
+tags: code quality culture
+---
+
+Here at Holiday Extras we've grown our web team from just a few developers to 40, with fingers on code across multiple repositories on a daily basis. Scaling code is hard, and unless you really fight for quality then mediocrity will creep in as a simple byproduct of so our inability to read minds - individual development skill aside.
```
Comment (Joezo) [inline _posts/2015-03-18-5-tips-for-a-smoother-code-review-process.md]: think you've left the word `so` in here `of so our inability`

File: _posts/2015-03-18-5-tips-for-a-smoother-code-review-process.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+At Holiday Extras, for example, we put things like _"I've checked there is appropriate unit test coverage."_, _"I've tested this on desktop IE9/10/11, Firefox, Chrome for any visual changes."_ and _"I've informed those who need to know of any unforeseen impact of this work"_ on the authors task list, and on the reviewers task list we have things like _"I’ve checked for appropriate test coverage.", _"I’ve checked for coding anti-patterns."_ and _"I don’t believe this work introduces unnecessary technical debt."_. You can roll these into your pull request template to ensure they're never forgotten.
+
+Use these task lists to give the proper accountability for the quality of work shipping. It's difficult to explain away a IE breaking change as an "I forgot to test it" if you've ticked a box telling everyone you did. It also means reviewers don't need to ask so many questions about what's been done prior to review, it's all just there for them to see right away.
```
Comment (Joezo) [inline _posts/2015-03-18-5-tips-for-a-smoother-code-review-process.md]: I had trouble comprehending this bit: `It's difficult to explain away a IE breaking change as` not sure if the `a` should be an `an` or whether there is a better way to word it...

File: _posts/2015-03-18-5-tips-for-a-smoother-code-review-process.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+tags: code quality culture
+---
+
+Here at Holiday Extras we've grown our web team from just a few developers to 40, with fingers on code across multiple repositories on a daily basis. Scaling code is hard, and unless you really fight for quality then mediocrity will creep in as a simple byproduct of so our inability to read minds - individual development skill aside.
+
+Code review has made a huge difference in our fight for better software, but it's hard to do well, and is trickier to do with larger teams and larger changes without introducing a bottleneck that prevents frequent deploys to production. So, from lessons we've learned the hard way, here are 5 tips for a smoother code review process.
+
+___
+
+## 1. Use and enforce pull pull request templates
```
Comment (lukefrake) [inline _posts/2015-03-18-5-tips-for-a-smoother-code-review-process.md]: pull pull

File: _posts/2015-03-18-5-tips-for-a-smoother-code-review-process.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+published: true
+title: "5 tips for a smoother code review process"
+author: mike_carter
+categories: ""
+tags: code quality culture
+---
+
+Here at Holiday Extras we've grown our web team from just a few developers to 40, with fingers on code across multiple repositories on a daily basis. Scaling code is hard, and unless you really fight for quality then mediocrity will creep in as a simple byproduct of so our inability to read minds - individual development skill aside.
+
+Code review has made a huge difference in our fight for better software, but it's hard to do well, and is trickier to do with larger teams and larger changes without introducing a bottleneck that prevents frequent deploys to production. So, from lessons we've learned the hard way, here are 5 tips for a smoother code review process.
```
Comment (lukefrake) [inline _posts/2015-03-18-5-tips-for-a-smoother-code-review-process.md]: 'Code reviews have made' > 'Code review has made'

Comment (lukefrake) [inline _posts/2015-03-18-5-tips-for-a-smoother-code-review-process.md]: 'From lessons learned the hard way' > 'from lessons we've learned the hard way' - Just opinion, nothing actually wrong with what you've put.

Comment (lukefrake) [inline _posts/2015-03-18-5-tips-for-a-smoother-code-review-process.md]: 'here are our top 5 tips' > 'here are 5 tips' - Just an opinion

**Comment by Joezo** [comment]
This is neat, i'll take the editor 1 spot. It would be "cool" if you linked to this PR in your blog post, so something about adapting it for blog posts etc then using that very post as the example. I like it though, left a couple of inline comments, will approve once they're sorted :)

File: _posts/2015-03-18-5-tips-for-a-smoother-code-review-process.md (hunk 4 above)
Comment (lukefrake) [inline _posts/2015-03-18-5-tips-for-a-smoother-code-review-process.md]: The first sentence in here is really long, I would split it up a bit if possible.

File: _posts/2015-03-18-5-tips-for-a-smoother-code-review-process.md [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+___
+
+## 1. Use and enforce pull pull request templates
+Create a default pull request template that all pull request authors must fill in before submitting their pull requests helps to ensure code is reviewed properly and the author is properly explaining the change they're making to reviewers.
```
Comment (lukefrake) [inline _posts/2015-03-18-5-tips-for-a-smoother-code-review-process.md]: Would it be good to introduce the term PR here? That way you don't have to repeat pull requst.

Maybe something like - Create a default PR (pull request) template?

File: _posts/2015-03-18-5-tips-for-a-smoother-code-review-process.md [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+___
+
+## 1. Use and enforce pull pull request templates
+Create a default pull request template that all pull request authors must fill in before submitting their pull requests helps to ensure code is reviewed properly and the author is properly explaining the change they're making to reviewers. 
+
+For us, these take the form of multiple headings explaining various aspects of the pull request, for example _"What does this PR do?"_, _"What tests does this PR have?"_, _"How should this work be manually tested?"_, _"What are the relevant tickets?"_. Forcing the author to fill these in thoroughly helps to give reviewers the context for the change before they get to the code step.
+
+For this to work, it's important to flat out reject pull requests that don't properly utilise your template without good reason. You can see the pull request template we use at Holiday Extras in our [hx culture repository](https://github.com/holidayextras/hx-culture).
```
Comment (lukefrake) [inline _posts/2015-03-18-5-tips-for-a-smoother-code-review-process.md]: The hx-culture page is currently private :-1: 

Also I would link directly to the PR template i.e. maybe something like this -
'We have open sourced our [pull request template](https://github.com/holidayextras/hx-culture/blob/master/pr-template.mkd) so feel free to use it as an example.'

File: _posts/2015-03-18-5-tips-for-a-smoother-code-review-process.md [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+___
+
+## 2. Use author and reviewer task lists
+If you're working with GitHub, the [task list](https://github.com/blog/1375%0A-task-lists-in-gfm-issues-pulls-comments) feature is absolutely fantastic for making sure the author and code reviewers have taken all the correct steps prior to merging work.
+
+At Holiday Extras, for example, we put things like _"I've checked there is appropriate unit test coverage."_, _"I've tested this on desktop IE9/10/11, Firefox, Chrome for any visual changes."_ and _"I've informed those who need to know of any unforeseen impact of this work"_ on the authors task list, and on the reviewers task list we have things like _"I’ve checked for appropriate test coverage.", _"I’ve checked for coding anti-patterns."_ and _"I don’t believe this work introduces unnecessary technical debt."_. You can roll these into your pull request template to ensure they're never forgotten.
```
Comment (lukefrake) [inline _posts/2015-03-18-5-tips-for-a-smoother-code-review-process.md]: I wouldn't use so many examples here as it makes it a little difficult to read & on any examples, can you drop the trailing full stop.

Comment (lukefrake) [inline _posts/2015-03-18-5-tips-for-a-smoother-code-review-process.md]: _"I’ve checked for appropriate test coverage." is missing it's trailing _

File: _posts/2015-03-18-5-tips-for-a-smoother-code-review-process.md (hunk 2 above)
Comment (lukefrake) [inline _posts/2015-03-18-5-tips-for-a-smoother-code-review-process.md]: :+1:

File: _posts/2015-03-18-5-tips-for-a-smoother-code-review-process.md [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+If you're a code author then solicit your reviewers directly 1 on 1, and ask them to commit to looking at your code before a certain time. Often it won't be convenient for them to review it right then and there, but if they commit to do it by a particular time it's much more likely to get done.
+
+There are plenty of other techniques to successfully recruiting reviewers. For example, the _"I'll review yours if you review mine"_, or the _"I can help you on that thing as soon as I get another review on my pull request"_ approaches work brilliantly.
+
+___
+
+## 4. Put your team's best practices down in writing
+Ever find yourself bringing up the same points over and over in code review across different pull requests? Put these down in a shared best practices document or repository of documents. This way, instead of rewriting the same comments over and over you can just link off to a markdown document that eloquently explains a particular antipattern and various potential improvements.
```
Comment (lukefrake) [inline _posts/2015-03-18-5-tips-for-a-smoother-code-review-process.md]: I would mention here about how linting helps reduce the things we repeat over and over again

File: _posts/2015-03-18-5-tips-for-a-smoother-code-review-process.md [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+Over time you'll find that when in doubt, people check the documents as a reference point for antipatterns and coding standards prior to putting in pull requests, and code quality will improve automatically.
+
+At Holiday Extras we use a separate [hx culture repository](https://github.com/holidayextras/hx-culture) for our best practice markdown docs. This way, people are free to challenge the best practices through pull requests of their own into the documents themselves.
```
Comment (lukefrake) [inline _posts/2015-03-18-5-tips-for-a-smoother-code-review-process.md]: You've repeated Holiday Extras 4 times, I don't think thats necessary and I think just 'We use a separate...' might sound nicer?

Maybe some of the other places you've said HX can be changed to just we? - Just an idea though :)

Comment (lukefrake) [inline _posts/2015-03-18-5-tips-for-a-smoother-code-review-process.md]: Again, hx culture private :-1:

File: _posts/2015-03-18-5-tips-for-a-smoother-code-review-process.md [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+
+___
+
+## 5. Group review large pull requests
+While we try to keep pull requests as small as possible, sometimes they do run to over a thousand changes. At this size, it's very difficult for individual code reviewers to get a high level view for how the change the author has submitted works. The code review step on pull requests like these takes forever, and often things get missed due to lack of understanding.
```
Comment (lukefrake) [inline _posts/2015-03-18-5-tips-for-a-smoother-code-review-process.md]: 'sometimes they do run over thousands of lines' > 'sometimes they do run to over a thousand changes' ?

Comment (lukefrake) [inline _posts/2015-03-18-5-tips-for-a-smoother-code-review-process.md]: Maybe good to reiterate at the end of this again that - breaking changes down > massive code reviews

File: _posts/2015-03-18-5-tips-for-a-smoother-code-review-process.md [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+An easy solution to this is for the author and reviewers to meet together in front of a TV or projector and run through the pull request diff up on a big screen. Lead by the author, this gives he or she the chance to explain their work to the reviewers, and the reviewers the chance to stop the author to ask questions or suggest improvements, validating their understanding every step of the way.
```
Comment (lukefrake) [inline _posts/2015-03-18-5-tips-for-a-smoother-code-review-process.md]: 'request diff up on a big screen, lead by the author. This gives' > 'request diff up on a big screen. Lead by the author, this gives'

Comment (lukefrake) [inline _posts/2015-03-18-5-tips-for-a-smoother-code-review-process.md]: 'the author' > 'he or she'

**Comment by lukefrake** [comment]
Lots of my points are opinions, so feel free to ignore as appropriate. Really good read mate!

**Comment by michaelcarter** [comment]
Working on getting that repo public, unfortunately our security policy is in there. We don't want to be sharing that.

**Comment by hpoom** [comment]
Move the security policy out to somewhere else, another private repo, then make it public.

**Comment by stuartstobie** [comment]
Looks like the review slots have all be taken up but this is really good :+1:

**Comment by lukefrake** [comment]
I don't think i can :+1: this post until the culture page is public, I don't think it makes sense to shout about a post where nobody can read the main part?

Can we move the security policy into a hx-private-docs? Pull the repo down and force push back up to remove history?

That said I had a quick look over the security policy, and I think that can be open sourced as well with a small tweak, maybe @hxoliverrumbelow can turn it into a blog post of 101 security basics or something?

**Comment by hpoom** [comment]
When is the blog post going out? I was trying to point people at a meetup tonight ot this blog post, but it is not yet published :disappointed: 

We need to get this out as I need to tweet it to about 5 people :pray:

**Comment by michaelcarter** [comment]
Hi, sorry. Just catching up with this now. I'm going to make a public gist of our PR template and link to that until I can get the culture repo public again. I need to know it's public and staying public before I link to it in blog posts unfortunately.

**Comment by michaelcarter** [comment]
@lukefrake Updated. :-)

**Comment by lukefrake** [comment]
Brilliant temporary solution! :+1: from me!

**Comment by stuartstobie** [comment]
+1'ing this in Joe's absence.

**Pull request merged**